napari/layers/graph/graph.py

```diff
@@ -0,0 +1,508 @@
+        is_buffer_domain : bool
+            Indicates if node indices are on world or buffer domain.
+        """
+        indices = np.atleast_1d(indices)
+        if indices.ndim > 1:
+            raise ValueError(
+                trans._(
+                    "Indices for removal must be 1-dim. Found {ndim}",
+                    ndim=indices.ndim,
+                )
+            )
+
+        prev_size = self.data.n_allocated_nodes
+
+        # it got error missing __iter__ attribute, but we guarantee by np.atleast_1d call
+        for idx in indices:  # type: ignore[union-attr]
+            self.data.remove_node(idx, is_buffer_domain)
+
+        self._data_changed(prev_size)
+
+    def _move_points(
+        self, ixgrid: Tuple[np.ndarray, np.ndarray], shift: np.ndarray
+    ) -> None:
+        """Move points along a set a coordinates given a shift.
+
+        Parameters
+        ----------
+        ixgrid : Tuple[np.ndarray, np.ndarray]
+            Crossproduct indexing grid of node indices and dimensions, see `np.ix_`
+        shift : np.ndarray
+            Selected coordinates shift
+        """
+        self.data.coords_buffer[ixgrid] = (
+            self.data.coords_buffer[ixgrid] + shift
+        )
+
+    def _update_props_and_style(self, data_size: int, prev_size: int) -> None:
+        # Add/remove property and style values based on the number of new points.
+        with self.events.blocker_all(), self._border.events.blocker_all(), self._face.events.blocker_all():
+            self._feature_table.resize(data_size)
+            self.text.apply(self.features)
+            if data_size < prev_size:
+                # If there are now fewer points, remove the size and colors of the
+                # extra ones
+                if len(self._border.colors) > data_size:
+                    self._border._remove(
+                        np.arange(data_size, len(self._border.colors))
+                    )
+                if len(self._face.colors) > data_size:
+                    self._face._remove(
+                        np.arange(data_size, len(self._face.colors))
+                    )
+                self._shown = self._shown[:data_size]
+                self._size = self._size[:data_size]
+                self._border_width = self._border_width[:data_size]
+                self._symbol = self._symbol[:data_size]
+
+            elif data_size > prev_size:
+                adding = data_size - prev_size
+
+                current_properties = self._feature_table.currents()
+                self._border._update_current_properties(current_properties)
+                self._border._add(n_colors=adding)
+                self._face._update_current_properties(current_properties)
+                self._face._add(n_colors=adding)
+
+                # `shown` must be first due to "refresh" calls inside `attribute`.setters
+                for attribute in ("shown", "size", "symbol", "border_width"):
```

Comment:
@JoOkuma @jni are we expecting to be able to add points to graphs in this implementation? If so, this line is problematic. 

Both `shown` and `size` setters have calls to `refresh`, which trigger slice updates. 

With `shown` first, we get stuck at indexing the `size` in [_get_out_of_slice_display_data](https://github.com/JoOkuma/napari/blob/napari-graph-2023/napari/layers/graph/_slice.py#L132), while with `size` first we get stuck at indexing `shown` in [_update_slice_response](https://github.com/JoOkuma/napari/blob/napari-graph-2023/napari/layers/points/points.py#L1508-L1509).

I feel like we need some sort of `refresh` blocker that can allow us to update the attributes before the view is refreshed. Thoughts?

Reply:
This sounds like a pretty deep architectural issue. I am thinking that refresh in setters is an anti-pattern: setters should trigger events, and then we can use the events blocker? CC @Czaki @andy-sweet who might have opinions/relevant experience about this...

Reply:
we did just strike upon `block_update_properties()` which seems to do something very similar for `points` and `shapes` layers. but we are not fully sure of its scope or how it works. we will continue to look into it, but any context/insight wrt the blocker would be amazing

Reply:
Given napari's current architecture of layers storing their slices, I think both solutions suggested here are viable for this PR.

Adding a similar refresh blocker to `block_update_properties` is almost certainly the easiest option. And lots of other layers call `refresh` in their property setters, so this would make the `Graph` layer consistent with those.  Here's an implementation that could be added to napari's base `Layer` and then could be used here.

```python
    @contextmanager
    def _block_refresh(self):
        previous = self._block_refresh
        self._block_refresh = True
        try:
            yield
        finally:
            self._block_refresh = previous

    def refresh(self):
        if self._block_refresh:
             return
        ...
```

I like @jni's idea of avoiding the explicit refresh in the property setters and instead connecting these events to refresh (e.g. in `__init__`) because it adds some separation of the changes to the data model that the layer represents (i.e. the side-effect of updating the slice in refresh is separated from the main effect of changing the property).

I think that approach has been avoided in the past in other layers because 1) those self-connections require some special treatment to avoid strong cyclic references and 2) it makes the program flow a little more in-direct and harder to follow. The other disadvantage is that the `Graph` layer would be inconsistent with most of the other layers - unless we change those too, which then adds more work.

My preferred long term solution here is to remove slice state from the layer. In that case the thing that owns the slices (e.g the vispy layer, or the canvas) can connect to the property events and handle refresh, which solves the problem here and many more. But doing this is clearly outside of the scope of this PR.

For this PR, I lean towards the refresh blocker as a quick workaround, because the property-setter/refresh connection seems like too much work that doesn't solve the core issue.

Reply:
Amazing, thanks so much for the detailed help. Looking into it.